### PR TITLE
Add handbook section for exposing-service

### PR DESCRIPTION
**Pull request by daxmc99**
Add a section on how to expose services

### CODE REVIEW

File: handbook/engineering/languages/exposing-services.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+http.ListenAndServe(":80", nil)
+```
+The above code will bind to all TCP interfaces. Since Kubernetes does support dual-stack IPv4 & IPv6 our services should bind to all interfaces (we do not currently have IPv6 services).
+
+If you must specify an ip address to expose your service on, choosing `0.0.0.0:port` is typically a good choice. What this does can be OS and platform-specific.
+
+Binding to `localhost:port` or `127.0.0.1:port` is binding to a local-only interface that constrained to the same "host". In Kubernetes, other containers within the Pod may still communicate with this service AND you may port-forward this container and associated service
+[(Why?)](#How-can-I-port-forward-a-local-only-service?).
+
+This may be preferred in a sidecar pattern where you do not want a container accessible outside a Pod or when you don't want expose your laptop to the cofeeshop wifi but generally, code should not be merged that binds to `localhost` or `127.0.0.1`.
```
Comment (bobheadxi) [inline handbook/engineering/languages/exposing-services.md:15]: Might want to point out that we have a CI check for this: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/dev/check/no-localhost-guard.sh?subtree=true

File: handbook/engineering/languages/exposing-services.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+
+### Exposing services
```
Comment (unknwon) [inline handbook/engineering/languages/exposing-services.md:2]: qq: why not 
```suggestion
# Exposing services
```

File: handbook/engineering/languages/exposing-services.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (unknwon) [inline handbook/engineering/languages/exposing-services.md:1]:
```suggestion
```

File: handbook/engineering/languages/go.md [hunk 4]
```diff
@@ -6,6 +6,10 @@ For all things not covered in this document, defer to
 [Go Code Review Comments](https://code.google.com/p/go-wiki/wiki/CodeReviewComments)
 and [Effective Go](http://golang.org/doc/effective_go.html).
 
+We also have subsections here
```
Comment (unknwon) [inline handbook/engineering/languages/go.md:9]:
```suggestion
We also have subsections here:
```

File: handbook/engineering/languages/exposing-services.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+### Exposing services
+
+In Go, that looks like this
```
Comment (unknwon) [inline handbook/engineering/languages/exposing-services.md:4]:
```suggestion
In Go, that looks like this:
```

File: handbook/engineering/languages/exposing-services.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+The above code will bind to all TCP interfaces. Since Kubernetes does support dual-stack IPv4 & IPv6 our services should bind to all interfaces (we do not currently have IPv6 services).
+
+If you must specify an ip address to expose your service on, choosing `0.0.0.0:port` is typically a good choice. What this does can be OS and platform-specific.
+
+Binding to `localhost:port` or `127.0.0.1:port` is binding to a local-only interface that constrained to the same "host". In Kubernetes, other containers within the Pod may still communicate with this service AND you may port-forward this container and associated service
+[(Why?)](#How-can-I-port-forward-a-local-only-service?).
+
+This may be preferred in a sidecar pattern where you do not want a container accessible outside a Pod or when you don't want expose your laptop to the cofeeshop wifi but generally, code should not be merged that binds to `localhost` or `127.0.0.1`.
+
+
```
Comment (unknwon) [inline handbook/engineering/languages/exposing-services.md:17]:
```suggestion
```

File: handbook/engineering/languages/exposing-services.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+You can also use `kubectl port-forward --address 0.0.0.0` if you need to expose a port-forwarded service outside of your local machine (it defaults to `127.0.0.1`). [link](https://github.com/kubernetes/kubernetes/issues/40053) :exploding_head:
+
+### References
+
+https://stackoverflow.com/questions/20778771/what-is-the-difference-between-0-0-0-0-127-0-0-1-and-localhost
+
+https://serverfault.com/questions/21657/semantics-of-and-0-0-0-0-in-dual-stack-oses/39561#39561
+
+https://stackoverflow.com/questions/49067160/what-is-the-difference-in-listening-on-0-0-0-080-and-80
```
Comment (unknwon) [inline handbook/engineering/languages/exposing-services.md:30]:
```suggestion
- https://stackoverflow.com/questions/20778771/what-is-the-difference-between-0-0-0-0-127-0-0-1-and-localhost
- https://serverfault.com/questions/21657/semantics-of-and-0-0-0-0-in-dual-stack-oses/39561#39561
- https://stackoverflow.com/questions/49067160/what-is-the-difference-in-listening-on-0-0-0-080-and-80
```

File: handbook/engineering/languages/exposing-services.md (hunk 1 above)
Comment (daxmc99) [reply handbook/engineering/languages/exposing-services.md:14]: Hmm... why did this not fail this PR then?...
https://github.com/sourcegraph/sourcegraph/pull/12758/files#r466173773
.................. 🙈

**Comment by daxmc99** [comment]
😬 Thanks @unknwon, I am so used to go fmt.

File: handbook/engineering/languages/exposing-services.md (hunk 1 above)
Comment (bobheadxi) [reply handbook/engineering/languages/exposing-services.md:14]: Yeah not too sure about the context of that check in particular, but I've ran into it failing my build before and it's a related topic so 🤷

**Pull request merged**